File: src/backend/envio-datos.py
from flask import Flask, request, jsonify
from flask_cors import CORS
import openai
import os

app = Flask(__name__)
CORS(app)
openai.api_key = os.getenv("OPENAI_API_KEY")

NORMAS_CHATGPT = """
Tienes que crear una imagen de un roadmap de lo que te pida el usuario siguiendo las indicaciones que te de. La imagen debe ser en formato PNG y tener un estilo visual minimalista con cajas de texto rectangulares de color azul claro y flechas de color gris oscuro conectándolas. El diseño del roadmap debe ser horizontal, mostrando las fases principales de izquierda a derecha.
Además, tienes que darme una breve descripción de la imagen que has creado (máximo tres frases) resaltando la estructura general del roadmap. También debes proporcionar un resumen del roadmap en formato de lista con viñetas, indicando los principales hitos y sus correspondientes acciones. Finalmente, indica si algún paso del roadmap requiere conocimientos técnicos específicos, mencionando brevemente cuáles.
"""
@app.route('/recogidaDatos', methods=['POST'])
def recogida_datos():
    data = request.get_json()
    prompt = data.get('prompt', '')
    checkbox = data.get('checkboxes', True)

    if not prompt:
        return jsonify({'error': 'No input provided'}), 400

    promptConNormas = f"{NORMAS_CHATGPT}\n\nPrompt del usuario: {prompt}"

    try:
        response = openai.ChatCompletion.create(
            model="gpt-3.5-turbo",
            messages=[
                {"role": "user", "content": promptConNormas}
            ],
            temperature=0.7,
            max_tokens=1500,
            n=1,
            stop=None
        )
        respuesta_chatgpt = response.choices[0].message.content
        return jsonify({'respuesta': respuesta_chatgpt})
    except Exception as e:
        print(f"Error al comunicarse con OpenAI: {e}")
        return jsonify({'error': 'Error al procesar la solicitud'}), 500

if __name__ == '__main__':
    app.run(debug=True)

File: src/backend/envio-datos.js
require('dotenv').config();
const express = require('express');
const cors = require('cors');
const { OpenAI } = require('openai');

const app = express();
app.use(cors());
app.use(express.json());

const openai = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY,
});

const NORMAS_CHATGPT = `
Eres un asistente que genera roadmaps visuales. Cuando recibas un prompt, debes:
    1. Crear una descripción de una imagen que represente el roadmap.
    2. Generar un prompt adecuado para DALL·E-3 que permita crear esa imagen empieza con "Prompt para la imagen: ".
    3. Proporcionar una breve descripción textual del roadmap.
    4. Crear un resumen con viñetas de los hitos del roadmap.
    5. Indicar si algún paso requiere conocimientos técnicos, y cuáles.
`;

app.post('/recogidaDatos', async (req, res) => {
    const { prompt } = req.body;
    console.log("ENVIO-DATOS ESTA FUNCIONANDO");
    if (!prompt) {
        return res.status(400).json({ error: 'No input provided' });
    }

    const prompt_con_normas = `${NORMAS_CHATGPT}\n\nPrompt del usuario: ${prompt}`;

    try {
        // Paso 1: generar respuesta de GPT-4
        const chatResponse = await openai.chat.completions.create({
        model: "gpt-4",
        messages: [
            { role: "user", content: prompt_con_normas }
        ],
        temperature: 0.7,
        max_tokens: 1500,
        });

        const respuestaChat = chatResponse.choices[0].message.content;

        // Buscar el prompt visual dentro de la respuesta
        const lines = respuestaChat.split('\n');
        const lineaPrompt = lines.find(line => line.toLowerCase().includes("prompt para la imagen"));
        const promptVisual = lineaPrompt ? lineaPrompt.split(':').slice(1).join(':').trim() : null;

        if (!promptVisual) {
        return res.status(500).json({ error: 'No se encontró prompt visual' });
        }

        // Paso 2: generar imagen con DALL·E-3
        const imageResponse = await openai.images.generate({
        model: "dall-e-3",
        prompt: promptVisual,
        size: "1024x1024",
        n: 1,
        });

        const imageUrl = imageResponse.data[0].url;

        return res.json({
        respuesta: respuestaChat,
        imagen: imageUrl,
        });

    } catch (error) {
        console.error("Error al comunicarse con OpenAI:", error);
        return res.status(500).json({ error: 'Error al procesar la solicitud' });
    }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
    console.log(`Servidor corriendo en http://localhost:${PORT}`);
});
